server/src/main/java/org/elasticsearch/cluster/metadata/DataLifecycle.java

```diff
@@ -53,31 +57,43 @@ public class DataLifecycle implements SimpleDiffable<DataLifecycle>, ToXContentO
     public static final ConstructingObjectParser<DataLifecycle, Void> PARSER = new ConstructingObjectParser<>(
         "lifecycle",
         false,
-        (args, unused) -> new DataLifecycle((TimeValue) args[0])
+        (args, unused) -> new DataLifecycle((Retention) args[0], false)
     );
 
     static {
-        PARSER.declareField(
-            ConstructingObjectParser.optionalConstructorArg(),
-            (p, c) -> TimeValue.parseTimeValue(p.textOrNull(), DATA_RETENTION_FIELD.getPreferredName()),
-            DATA_RETENTION_FIELD,
-            ObjectParser.ValueType.STRING_OR_NULL
-        );
+        PARSER.declareField(ConstructingObjectParser.optionalConstructorArg(), (p, c) -> {
+            String value = p.textOrNull();
+            if (value == null) {
+                return Retention.NULL;
+            } else {
+                return new Retention(TimeValue.parseTimeValue(value, DATA_RETENTION_FIELD.getPreferredName()));
+            }
+        }, DATA_RETENTION_FIELD, ObjectParser.ValueType.STRING_OR_NULL);
     }
 
     public static boolean isEnabled() {
         return DLM_FEATURE_FLAG.isEnabled();
     }
 
     @Nullable
-    private final TimeValue dataRetention;
+    private final Retention dataRetention;
+    private final boolean nullified;
```

Comment:
I think it's probably worth a comment here explaining what `nullified` is, and what it means if `dataRetention` is `null` vs if `dataRetention` is `Retention.NULL`. I know it's fairly straightforward, but I keep tripping up over it and I'm pretty familiar with this code at this point. It might save some time for someone coming to this without any context (or even one of us in a couple of years)

Reply:
You are right, there are subtle differences that shouldn't be ignored.

Reply:
I think this kind of logic shouldn't be in the cluster state model. 

I think this field shouldn't exist in the `DataLifecycle` as it seems to be a marker that has to do with the parsing of the `template` and to the composability of lifecycle (which belongs in the index/component templates space). 

I think we can indicate that `lifecycle: null` was explicitly configured when parsing the `Template` object using either:
- a subclass of `DataLifecycle` 
e.g. in `Template.java`:
 ```
    private static final class NullLifecycle extends DataLifecycle {
        static final NullLifecycle INSTANCE = new NullLifecycle(); 
    }
```
and in the `static` declaration where we define the `PARSER` we could do:
```
        if (DataLifecycle.isEnabled()) {
            PARSER.declareObjectOrNull(ConstructingObjectParser.optionalConstructorArg(), (p, c) -> DataLifecycle.fromXContent(p),
                NullLifecycle.INSTANCE, LIFECYCLE);
        }
    }
```

- a marker instance of the `DataLifecycle` itself - e.g. 
e.g. in `Template.java`:
```
public static final DataLifecycle NULL_LIFECYCLE_INSTANCE = new DataLifecycle();
```
and in the `static` declaration where we define the `PARSER` we could do:
```
        if (DataLifecycle.isEnabled()) {
            PARSER.declareObjectOrNull(ConstructingObjectParser.optionalConstructorArg(), (p, c) -> DataLifecycle.fromXContent(p),
                Template.NULL_LIFECYCLE_INSTANCE, LIFECYCLE);
        }
    }
```

I think the `compose` method could also be something that we move to `MetadataIndexTemplateService` so `DataLifecycle` is agnostic of how it's used (and will not know of `NullLifecycle` or the special instance to indicate the user requested a `null` lifecycle - whichever option we end up using)

What do you think @gmarouli ?

Reply:
From a first glance I really like it, it addresses a lot of "uncomfortable" parts of the code. Let me give it a try and see if it turns out as good as I think it will! Thank you @andreidan!